crates/miden-block-prover/src/local_block_prover.rs

```diff
@@ -0,0 +1,301 @@
+        let proof_hash = Digest::default();
+
+        let header = BlockHeader::new(
+            version,
+            prev_block_commitment,
+            block_num,
+            new_chain_root,
+            new_account_root,
+            new_nullifier_root,
+            note_root,
+            tx_hash,
+            kernel_root,
+            proof_hash,
+            timestamp,
+        );
+
+        // Construct the new proven block.
+        // --------------------------------------------------------------------------------------------
+
+        let proven_block = ProvenBlock::new_unchecked(
+            header,
+            updated_accounts,
+            output_note_batches,
+            created_nullifiers,
+        );
+
+        Ok(proven_block)
+    }
+}
+
+/// Computes the new nullifier root by inserting the nullifier witnesses into a partial nullifier
+/// tree and marking each nullifier as spent in the given block number. Returns the list of
+/// nullifiers and the new nullifier tree root.
+fn compute_nullifiers(
+    created_nullifiers: BTreeMap<Nullifier, NullifierWitness>,
+    prev_block_header: &BlockHeader,
+    block_num: BlockNumber,
+) -> Result<(Vec<Nullifier>, Digest), ProvenBlockError> {
+    let nullifiers: Vec<Nullifier> = created_nullifiers.keys().copied().collect();
+
+    let mut partial_nullifier_tree = PartialNullifierTree::new();
+
+    // First, reconstruct the current nullifier tree with the merkle paths of the nullifiers we want
+    // to update.
+    // Due to the guarantees of ProposedBlock we can safely assume that each nullifier is mapped to
+    // its corresponding nullifier witness, so we don't have to check again whether they match.
+    for witness in created_nullifiers.into_values() {
+        partial_nullifier_tree
+            .add_nullifier_witness(witness)
+            .map_err(ProvenBlockError::NullifierWitnessRootMismatch)?;
+    }
+
+    // Check the nullifier tree root in the previous block header matches the reconstructed tree's
+    // root.
+    if prev_block_header.nullifier_root() != partial_nullifier_tree.root() {
+        return Err(ProvenBlockError::StaleNullifierTreeRoot {
+            prev_block_nullifier_root: prev_block_header.nullifier_root(),
+            stale_nullifier_root: partial_nullifier_tree.root(),
+        });
+    }
+
+    // Second, mark each nullifier as spent in the tree. Note that checking whether each nullifier
+    // is unspent is checked as part of the proposed block.
+    for nullifier in nullifiers.iter().copied() {
+        // SAFETY: As mentioned above, we can safely assume that each nullifier's witness was added
+        // and every nullifier should be tracked by the partial tree and therefore updatable.
+        partial_nullifier_tree.mark_spent(nullifier, block_num).expect(
+            "nullifier's merkle path should have been added to the partial tree and the nullifier should be unspent",
+        );
+    }
```

Comment:
I would maybe move this loop into a method on the `PartialNullifierTree` which would take a block num and an iterator over nullifiers to perform a "bulk update". The main reasoning for this is that at some point we will probably need to change this to a multi-threaded update and then the changes would be more "self-contained".